File: ex10/Atividade.java
// ============================================================
// ABSTRAÇÕES DOS PRODUTOS
// NÃO MODIFICAR, POR FAVOR
// ============================================================

interface Veiculo {
    String nome();
    String operar();
}


interface Sensor {
    String nome();
    String coletar();
}


interface Relatorio {
    String gerar(String resumo);
}


// ============================================================
// PRODUTOS CONCRETOS
// NÃO MODIFICAR, POR FAVOR
// ============================================================

class RoverDeserto implements Veiculo {

    @Override
    public String nome() {
        return "RoverDeserto";
    }

    @Override
    public String operar() {
        return "exploracao no deserto";
    }
}


class RoverPolar implements Veiculo {

    @Override
    public String nome() {
        return "RoverPolar";
    }

    @Override
    public String operar() {
        return "exploracao polar";
    }
}


class SensorTermico implements Sensor {

    @Override
    public String nome() {
        return "SensorTermico";
    }

    @Override
    public String coletar() {
        return "temperatura";
    }
}


class SensorGelo implements Sensor {

    @Override
    public String nome() {
        return "SensorGelo";
    }

    @Override
    public String coletar() {
        return "espessura do gelo";
    }
}


class RelatorioTexto implements Relatorio {

    @Override
    public String gerar(String resumo) {
        return "TEXTO: " + resumo;
    }
}


class RelatorioDetalhado implements Relatorio {

    @Override
    public String gerar(String resumo) {
        return "DETALHADO: " + resumo;
    }
}


// ============================================================
// ABSTRACT FACTORY
//
// Define o contrato para criação de uma família completa
// de equipamentos relacionados.
//
// NÃO MODIFICAR ESTA INTERFACE, POR FAVOR.
// ============================================================

interface EquipamentoFactory {

    Veiculo criarVeiculo();

    Sensor criarSensor();
}


// ============================================================
// SUA IMPLEMENTAÇÃO
//
// Complete as duas fábricas concretas.
// ============================================================

class DesertoFactory implements EquipamentoFactory {

    @Override
    public Veiculo criarVeiculo() {
        // TODO: implementar
        return null;
    }

    @Override
    public Sensor criarSensor() {
        // TODO: implementar
        return null;
    }
}


class PolarFactory implements EquipamentoFactory {

    @Override
    public Veiculo criarVeiculo() {
        // TODO: implementar
        return null;
    }

    @Override
    public Sensor criarSensor() {
        // TODO: implementar
        return null;
    }
}


// ============================================================
// FACTORY METHOD
//
// O fluxo geral da missão já está implementado.
// As subclasses decidem qual relatório será criado.
//
// NÃO MODIFICAR executar().
// ============================================================

abstract class Missao {

    public abstract Relatorio criarRelatorio();

    public final String executar(
            Veiculo veiculo,
            Sensor sensor
    ) {

        String resumo =
                veiculo.nome()
                + " | "
                + veiculo.operar()
                + " | "
                + sensor.nome()
                + " | "
                + sensor.coletar();

        Relatorio relatorio =
                criarRelatorio();

        return relatorio.gerar(resumo);
    }
}


// ============================================================
// SUA IMPLEMENTAÇÃO
//
// Complete o Factory Method das duas subclasses.
// ============================================================

class MissaoRapida extends Missao {

    @Override
    public Relatorio criarRelatorio() {
        // TODO: implementar
        return null;
    }
}


class MissaoCientifica extends Missao {

    @Override
    public Relatorio criarRelatorio() {
        // TODO: implementar
        return null;
    }
}


// ============================================================
// INTEGRAÇÃO
//
// O planejador deve trabalhar apenas com as abstrações
// Missao e EquipamentoFactory.
// ============================================================

class PlanejadorMissoes {

    public String executarMissao(
            Missao missao,
            EquipamentoFactory fabrica
    ) {

        // TODO:
        // 1. solicitar um veículo à fábrica;
        // 2. solicitar um sensor à fábrica;
        // 3. executar a missão com os objetos criados;
        // 4. retornar o resultado da execução.

        return null;
    }
}


// ============================================================
// Essa classe pública é necessária para a entrega como Atividade.java.
//
// Não é necessário implementar main.
// ============================================================

public class Atividade {
}

File: ex10/atividade.py
from abc import ABC, abstractmethod


# ============================================================
# ABSTRAÇÕES DOS PRODUTOS
# NÃO MODIFICAR, POR FAVOR
# ============================================================

class Veiculo(ABC):
    @abstractmethod
    def nome(self):
        pass

    @abstractmethod
    def operar(self):
        pass


class Sensor(ABC):
    @abstractmethod
    def nome(self):
        pass

    @abstractmethod
    def coletar(self):
        pass


class Relatorio(ABC):
    @abstractmethod
    def gerar(self, resumo):
        pass


# ============================================================
# PRODUTOS CONCRETOS
# NÃO MODIFICAR, POR FAVOR
# ============================================================

class RoverDeserto(Veiculo):
    def nome(self):
        return "RoverDeserto"

    def operar(self):
        return "exploracao no deserto"


class RoverPolar(Veiculo):
    def nome(self):
        return "RoverPolar"

    def operar(self):
        return "exploracao polar"


class SensorTermico(Sensor):
    def nome(self):
        return "SensorTermico"

    def coletar(self):
        return "temperatura"


class SensorGelo(Sensor):
    def nome(self):
        return "SensorGelo"

    def coletar(self):
        return "espessura do gelo"


class RelatorioTexto(Relatorio):
    def gerar(self, resumo):
        return "TEXTO: " + resumo


class RelatorioDetalhado(Relatorio):
    def gerar(self, resumo):
        return "DETALHADO: " + resumo


# ============================================================
# ABSTRACT FACTORY
#
# Define o contrato para criação de uma família de
# equipamentos relacionados.
#
# NÃO MODIFICAR ESTA CLASSE, POR FAVOR.
# ============================================================

class EquipamentoFactory(ABC):

    @abstractmethod
    def criar_veiculo(self):
        pass

    @abstractmethod
    def criar_sensor(self):
        pass


# ============================================================
# SUA IMPLEMENTAÇÃO
#
# Complete as duas fábricas concretas.
# ============================================================

class DesertoFactory(EquipamentoFactory):

    def criar_veiculo(self):
        return RoverDeserto()

    def criar_sensor(self):
        return SensorTermico()


class PolarFactory(EquipamentoFactory):

    def criar_veiculo(self):
        return RoverPolar()

    def criar_sensor(self):
        return SensorGelo()


# ============================================================
# FACTORY METHOD
#
# O fluxo geral da missão já está implementado.
# As subclasses devem decidir qual relatório será criado
# por meio de criar_relatorio().
#
# NÃO MODIFICAR executar(), POR FAVOR.
# ============================================================

class Missao(ABC):

    @abstractmethod
    def criar_relatorio(self):
        pass

    def executar(self, veiculo, sensor):

        resumo = (
            veiculo.nome()
            + " | "
            + veiculo.operar()
            + " | "
            + sensor.nome()
            + " | "
            + sensor.coletar()
        )

        relatorio = self.criar_relatorio()

        return relatorio.gerar(resumo)


# ============================================================
# SUA IMPLEMENTAÇÃO
#
# Complete o Factory Method de cada tipo de missão.
# ============================================================

class MissaoRapida(Missao):

    def criar_relatorio(self):
        return RelatorioTexto()


class MissaoCientifica(Missao):

    def criar_relatorio(self):
        return RelatorioDetalhado()


# ============================================================
# INTEGRAÇÃO
#
# O planejador deve trabalhar apenas com as abstrações
# Missao e EquipamentoFactory.
# ============================================================

class PlanejadorMissoes:

    def executar_missao(self, missao, fabrica):
        
        return missao.executar(fabrica.criar_veiculo(), fabrica.criar_sensor())